Make top bar accessible when zoomed in.

tellthemachines opened the pull request:
## Description
<!-- Please describe what you have changed or added -->

Partially fixes #15346 

Top bar should wrap at high zoom levels and small breakpoints. 
~Table block should scroll horizontally when its contents are wider than the screen, unless "fixed table width" is set. 
Words inside table cells should not break unless "fixed table width" is set.~

UPDATE: Reverted table changes on this PR and will do a separate PR for them.

## How has this been tested?
<!-- Please describe in detail how you tested your changes. -->
<!-- Include details of your testing environment, tests ran to see how -->
<!-- your change affects other areas of the code, etc. -->

Manually tested on Chrome/Mac, Safari and IE at 400% zoom; also checked on iPhone simulator.
Ran unit and e2e tests locally.

## Screenshots <!-- if applicable -->

<img width="796" alt="Screen Shot 2019-06-20 at 4 56 09 pm" src="https://user-images.githubusercontent.com/8096000/59828778-19883f80-937f-11e9-93c3-a20ed77f1c2d.png">

<img width="705" alt="Screen Shot 2019-06-20 at 4 55 26 pm" src="https://user-images.githubusercontent.com/8096000/59828787-1c833000-937f-11e9-9303-66ec05413b09.png">


## Types of changes
<!-- What types of changes does your code introduce?  -->
Bug fix (non-breaking change which fixes an issue)
<!-- New feature (non-breaking change which adds functionality) -->
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://wordpress.org/gutenberg/handbook/designers-developers/ -->

tellthemachines committed:
Make top bar and tables accessible when zoomed in.

tellthemachines committed:
Remove min-width for scrollability in IE

jasmussen commented on packages/block-library/src/table/editor.scss:
```diff
@@ -3,9 +3,14 @@
 	&[data-align="right"],
 	&[data-align="center"] {
 		table {
+			display: table;
```
I would like for us to get @ellatrix input on this one before we ship. I'm almost certain that I've explored something similar back in the past in efforts to make the table responsive, but every solution I found had some breaking downsides.

ellatrix replied:
I forgot the details as well. Isn't `table` the default `display` value in case of a `table` element? Why only for aligned tables? This should receive an inline comment.

tellthemachines replied:
Yup, sorry, I should have described the changes better. I'm setting the default display for the table to `block` so that, with `overflow-x: scroll`, it can scroll horizontally on small breakpoints, as recommended in the issue. However, for fixed-width tables and for aligned tables that won't work, so I'm resetting the display value to `table` for those. I'll put a comment in to make it clear.

tellthemachines replied:
Done!

afercia replied:
For future reference: it is important to note that [some CSS display properties do affect table semantics](http://adrianroselli.com/2018/02/tables-css-display-properties-and-aria.html#TheTable). For example, flex | block | grid | inline, they all override the table native semantics, rendering it useless to a screen reader. Keeping the native display: table is the right way to go for data tables, thanks for that! 🙂

jasmussen commented on packages/edit-post/src/components/header/style.scss:
```diff
@@ -1,19 +1,22 @@
-	align-items: stretch;
+	flex-flow: row wrap;
 	justify-content: space-between;
+	max-width: 100vw;
```
We should always be careful when using the vw unit, because it does not account for the presence of visible scrollbars. So using 100vw might unintentionally cause a horizontal scrollbar.

tellthemachines replied:
Good point. I checked Chrome and Safari on Mac, Chrome and IE on Windows 7 and Windows 10, and it's not happening, though likely that's because we have `overflow-x: hidden` on the body element for screens under 782px. If we removed that, we'd get a massive horizontal scroll because of the admin bar having a min-width of 300px. That, and a couple other min-widths set on the core side, are the reason I had to resort to `vw` in the first place, to get the top bar to not overflow the viewport.
I'll check trac to see if there's a ticket for addressing that, and create one if there isn't, but suspect it will be non-trivial to fix on that side 😅so for now we're probably better off with the `vw` solution.

jasmussen commented:
Wow, impressive work, thanks so much for thinking of, what in this case is _beyond mobile_. At a glance, the changes seem okay, if things things we want to both test very thoroughly.

I left a few comments in the code, but overally my primary question is how well this has been tested on non mobile, and very wide breakpoints. Due to the way the CSS and responsive mixins are structured, the CSS is written mobile first and then overridden on ever-increasing breakpoints. Which means the bulk of the CSS written here applies across every breakpoint, because there's nothing to "unset it". This may absolutely be intentional, but it also makes this PR a little more sensitive because it touches the layout of every breakpoint out there, necessitating a need for quite a bit of browser testing.

tellthemachines commented:
Thanks for the feedback @jasmussen! I have now checked the large/very large breakpoints on both Mac and Windows (on an actual laptop running Windows 10, and on a virtual machine running Windows 7), across Chrome, Firefox, Safari and IE, and also iPad pro with Safari on the Xcode simulator, and they're all displaying fine.

tellthemachines committed:
Add explanatory comment for display reset.

kjellr commented:
Hey @tellthemachines! In general, this looks excellent. Thank you for diving into this. 

I tested on a Mac, with scrollbars set to "Always", and have a little bit of feedback. 

When a table is brand new, it used to take up the entire width like this: 

<img width="620" alt="Screen Shot 2019-06-25 at 3 22 34 PM" src="https://user-images.githubusercontent.com/1202812/60126979-29c46400-975d-11e9-8344-29b0c3d09eec.png">

Is there any way we can preserve that? The PR currently presents me with this: 


Also, I see disabled scrollbars even when there's no content in the table yet:

tellthemachines commented:
Oh, good spot @kjellr! The scrollbar can be fixed by setting overflow to `auto` instead of `scroll`. The empty cells not taking up full width is due to the display on the table element being set to `block` to allow scrolling, and that doesn't play well with the default display modes of the children. I'll investigate if there is any CSS-only workaround for this, otherwise we might have to go back to using a `div` wrapper for the table block. That's a larger piece of work though, and I was hoping we could get away without any markup changes. 😅

tellthemachines commented:
Update on this: I can't find a way of making the empty table go full width with only CSS, so am going to go the wrapper div route. I have another branch with most of that work done already so will remove the table styles from this PR and do a separate one for the div wrapper. This PR should then only address the top bar issue.

tellthemachines committed:
Revert table changes

jasmussen commented on packages/edit-post/src/components/header/style.scss:
```diff
@@ -1,19 +1,22 @@
-	top: 0;
-	position: sticky;
+	// Make toolbar sticky on larger breakpoints
+	@include break-mobile {
```
Is there a reason you removed the sticky feature from mobile? In doing so, the publish and preview buttons require you to swipe all the way to the top to do so. Before:

![before-scrolling](https://user-images.githubusercontent.com/1204802/60247115-5218a380-98c0-11e9-8ff2-f2699353aef3.gif)

After:

![after-scrolling](https://user-images.githubusercontent.com/1204802/60247119-547afd80-98c0-11e9-9b4c-2f705eb00ed4.gif)

Changing this behavior is not off the table — a cogent argument can be made that there is insufficient editing space especially when the soft keyboard is showing. However I would think this change needs discussion separately, unless it is absolutely critical in order to make this PR work.

tellthemachines replied:
The reason I removed the sticky behaviour from mobile is that it was explicitly flagged in issue #15346 as a problem to be solved:

> For the post header set the position attribute to block instead
of sticky when a media query detects the viewport is small (20-30
em).

Design feedback was requested on that issue in early May and @mapk approved the proposed solution, which included the quote above. So I assumed it was safe to implement this solution :-)

For context, the 20-30 em mentioned should give us 280-420px at our body font size of 14. That is a bit smaller than our smallest breakpoint, which is 480px, but I was reluctant to add a new breakpoint. 

It may be justified to do so though, to cater explicitly to large zoom levels, so I propose adding a 280px breakpoint, which is more or less equivalent to a 640px screen width at 250% zoom. That's the point beyond which the top bar elements start to wrap.

jasmussen replied:
Thank you for resurfacing that for me, I missed this. 

Given that context, I think that a new "micro" 280px seems like a clever "best of both worlds" solution. If you feel this PR can handle it, that could work, otherwise a separate followup PR can do it.

tellthemachines replied:
Done!

jasmussen reviewed (dismissed):
Thank you for your continued work on this. 

I know it's just a few lines of code, but they also happen to be fundamental and structural to the entire layout. Unfortunately a few things broke in the process, that we'll have to fix first. 

The header bar is now taller, and items inside misaligned. Before:

![before](https://user-images.githubusercontent.com/1204802/60247503-16320e00-98c1-11e9-8084-253b2f6ba9bf.png)

After:

![after](https://user-images.githubusercontent.com/1204802/60247511-18946800-98c1-11e9-940b-1006ac81357c.png)

I left a comment in context of the code about the change to position: sticky; for mobile. As noted, this is a topic that can be discussed, but if it isn't crucially necessary for this PR, I would suggest exploring it separately, as it's a big change.

The columns block broken entirely. This branch:

![Screenshot 2019-06-27 at 09 47 57](https://user-images.githubusercontent.com/1204802/60247619-44afe900-98c1-11e9-86a2-4d9c3a31dd4d.png)

Master:

![Screenshot 2019-06-27 at 09 49 08](https://user-images.githubusercontent.com/1204802/60247634-4b3e6080-98c1-11e9-8266-d59e33d0d7d7.png)

Fullwide alignments broke. This branch:

![Screenshot 2019-06-27 at 09 48 08](https://user-images.githubusercontent.com/1204802/60247660-585b4f80-98c1-11e9-8eef-865ac992b548.png)

Master:

![Screenshot 2019-06-27 at 09 48 52](https://user-images.githubusercontent.com/1204802/60247673-5db89a00-98c1-11e9-9ffc-db3331eb5552.png)

I'm almost certain it's the vw unit width, which is not aware of neither left hand navigation sidebars, right hand settings sidebars, or even scrollbars in either of the three. If there is any way at all we can avoid the vw unit, that would be preferable as it can theoretically be the same code across all breakpoints and zoom levels leading to better code quality. If the vw unit truly is the only way forward, it needs to be unset beyond the mobile breakpoint.

I understand this is a priority and I deeply appreciate your work here, but I believe the best way forward is to change as little code as possible, in as small a PR as possible. I know it's already pretty small, but some things to try:

- Use min-height instead of removing the height and moving it to the mobile+ breakpoint.
- Instead of using flex shorthand, keep the specific flex properties as they are, but remove or add any non-shorthand additional flex properties as needed.
- Avoid vw if at all possible. If 100% works, that will likely cause fewer cascading effects.
- Avoid changing the sticky unless it's necessary. 

Thanks again.

tellthemachines committed:
Re-align top bar elements.

tellthemachines commented:
The misalignment of the top bar was well spotted @jasmussen ! That was actually due to the `flex-wrap` I set on `edit-post-header__settings`. I have fixed it now. 

Regarding the issue with the horizontal scroll on the full width image, that was an unrelated regression and was fixed in #16295. I have rebased master so the fix should now be on this branch.

The columns issue is a funny one. Before rebasing, I was seeing the previous version of the block, where creating a new columns block immediately brings up a default two-column layout:

<img width="993" alt="Screen Shot 2019-06-28 at 12 50 26 pm" src="https://user-images.githubusercontent.com/8096000/60320342-b1f47480-99bc-11e9-9ce9-d15bd0511a26.png">

After rebasing, I now see the placeholder displaying correctly on my branch, but I was able to reproduce the issue on your screenshot by clicking the sidebar. That behaviour is unrelated to my changes though,  and I can reproduce it on master, so I have created issue #16343 for it.

jasmussen commented:
Gosh, it sounds like a lot of those issues _were_ indeed rebase related. Thanks yet again for your continued work on this.

I am about to be AFK for the day, so I can't verify everything works as intended. I also maintain that we should be exceptionally careful about using the vw unit, and if you can use percent instead, it would very probably be more resilient code for it. Did you have any thoughts on this?

If someone else has the time to code review this and approve it before I'm able to test, they should feel free to dismiss <a href="https://github.com/WordPress/gutenberg/pull/16250#pullrequestreview-255060980">my change requests if they are addressed</a>.

tellthemachines commented:
Haha @jasmussen I'm just now replying to your question on my other PR about the `vw` issue. No worries, hopefully we can sort this out soon 😄

tellthemachines committed:
Unstick top bar only when zoomed in.

jasmussen commented:
In testing this, everything seems to work as the PR intends. None of the previously mentioned issues are present anymore. 

The editing canvas isn't usable at all at this zoom level:

<img width="566" alt="Screenshot 2019-07-02 at 10 08 17" src="https://user-images.githubusercontent.com/1204802/60495738-b026fb80-9cb1-11e9-8cab-9ca50090629f.png">

... but I imagine that is separate from this PR, especially given the offending overflow-x is previous core code. 

The vw unit is still added as a max-width, which as I've noted in the other PR feels like the wrong approach. But our very level-headed friend, @kjellr gave some great thoughts on https://github.com/WordPress/gutenberg/pull/16255#discussion_r299153871, and I imagine the same thoughts apply here. So I've dismissed my own change request, and if Kjell can confirm the same thoughts apply here, we can merge this right in. Thanks again for your work.

kjellr commented on assets/stylesheets/_breakpoints.scss:
```diff
@@ -10,6 +10,7 @@ $break-large: 960px;	// admin sidebar auto folds
 $break-medium: 782px;	// adminbar goes big
 $break-small: 600px;
 $break-mobile: 480px;
+$break-zoomedin: 280px;
```
This variable name reads a little awkwardly to me — should it be`break-min-zoom` or something? Or perhaps we just need to add a dash in, so that it reads as `break-zoomed-in`. Interested in what @jasmussen thinks.

jasmussen replied:
Great thoughts. Both your suggestions are nice.

tellthemachines replied:
Done! Going with `zoomed-in` so it's obvious what the use case scenario is.

kjellr commented:
> The vw unit is still added as a max-width, which as I've noted in the other PR feels like the wrong approach. But our very level-headed friend, @kjellr gave some great thoughts on #16255 (comment), and I imagine the same thoughts apply here. So I've dismissed my own change request, and if Kjell can confirm the same thoughts apply here, we can merge this right in. Thanks again for your work.

Agreed, I think this is another one of those circumstances. Like the other PR, we should add a code comment referencing the Trac issue. @tellthemachines, would you mind adding that in?

Aside from that and a minor variable name question I raised above, I think this is ready to go. It looks like the ellipsis menu is [basically unusable at super-high zoom levels](https://cloudup.com/ct0IWPMkNRR), but that's a different PR. 😄 

Thanks for the work here, @tellthemachines!

tellthemachines committed:
Address PR feedback.

kjellr reviewed (approved):
This is working well in my tests. Thanks for all the work on this, @tellthemachines!
